swap/src/asb/recovery/cancel.rs

```diff
@@ -42,8 +41,15 @@ pub async fn cancel(
         | AliceState::SafelyAborted => bail!("Swap is is in state {} which is not cancelable", state),
     };
 
-    // TODO: handle these error variants?
-    let txid = state3.submit_tx_cancel(bitcoin_wallet.as_ref()).await?;
+    let txid = match state3.submit_tx_cancel(bitcoin_wallet.as_ref()).await {
+        Ok(txid) => txid,
+        Err(err) => {
+            if let Some(bdk::Error::TransactionConfirmed) = err.downcast_ref::<bdk::Error>() {
```

Comment:
Are you sure we are actually hitting this variant? Aren't all the electrum errors crammed into `bdk::Error::Electrum(bdk::electrum_client::Error::Protocol(..))`?

Reply:
not sure. I think having the check was actually better. Downcasting and parsing the error seems fragile.  First I will have to figure out the string that corresponds and then check  `bdk::electrum_client::Error::Protocol(String)` if it matches. And also hope that it doesn't change across versions/implementations.

Reply:
Maybe I could try and submit tx_cancel. If it fails, I check if it has already been submitted and then log that "cancel has been published". This minimizes the amount of things that can go wrong before submission of tx cancel but is also robust when determining if the submission failed because tx cancel has already been submitted.

Reply:
To be perfectly honest, I think checking the error is better because it is an atomic operation. There is no guarantee that the state of the external system is the same between your two RPC calls. Assuming Electrum is implemented in a sane way, parsing the error will let you know why things didn't work. It is a reasonable assumption that the API will not undergo breaking changes!